### PR TITLE
Configuration-based PluginTestHarness instead of plugin object

### PR DESCRIPTION
**Pull request by SimonSapin**

### CODE REVIEW

File: apollo-router/src/plugin/mod.rs [hunk 1]
```diff
@@ -312,6 +320,16 @@ where
     }
 }
 
+impl dyn DynPlugin {
+    pub(crate) fn downcast_ref<T: 'static>(&self) -> Option<&T> {
+        if TypeId::of::<T>() == self.type_id() {
+            unsafe { Some(&*(self as *const dyn DynPlugin as *const T)) }
```
Comment (garypen) [inline apollo-router/src/plugin/mod.rs:326]:
```suggestion
            Some(unsafe { &*(self as *const _ as *const T) })
```

**Review by garypen — COMMENTED** [review]
I've speed-read through the changes and I think the direction is good.

I have got one question: How will a user know the valid configuration to supply for a plugin?

Apart from that, I really like the clean-up effect on the examples.

File: apollo-router/src/plugin/mod.rs (hunk 1 above)
Comment (SimonSapin) [reply apollo-router/src/plugin/mod.rs:326]: This line was copy-pasted from the standard library, but I don’t mind making that change if you prefer. I can also add a comment with the source URL

Comment (SimonSapin) [reply apollo-router/src/plugin/mod.rs:326]: I also tend to prefer not relying on inference too much when casting raw pointers, it’s easy to end up accidentally casting from `&&Foo` instead of `&Foo`

Comment (garypen) [reply apollo-router/src/plugin/mod.rs:326]: Good points. How about:
```
Some(unsafe { &*(self as *const dyn DynPlugin as *const T) })
```
?

**Comment by SimonSapin** [comment]
> How will a user know the valid configuration to supply for a plugin?

It is true that when a user writes a `json!( … )` macro they get less compile-time checks or IDE-provided guidance about the expected "shape" of plugin configuration, compared to a Rust struct literal. However tests for a plugin are typically written not far from the source code of that plugin so users can look there, and the new situation is no worse that using that plugin when writing a YAML router configuration file.

File: apollo-router/src/plugin/mod.rs (hunk 1 above)
Comment (SimonSapin) [reply apollo-router/src/plugin/mod.rs:326]: Done!

File: apollo-router/src/services/mod.rs [hunk 2]
```diff
@@ -122,8 +122,16 @@ impl RouterRequest {
         variables: HashMap<String, Value>,
         extensions: HashMap<String, Value>,
         context: Option<Context>,
-        headers: MultiMap<IntoHeaderName, IntoHeaderValue>,
+        mut headers: MultiMap<IntoHeaderName, IntoHeaderValue>,
```
Comment (garypen) [inline apollo-router/src/services/mod.rs:125]: This is a breaking change and should be noted in the changelog. I honestly don't think anyone will be affected by this change, but we should document it.

Something like:
With this change, if a request does not specify a `content-type`, then a default `content-type` of `application/json` will be provided.

File: apollo-router/src/services/execution_service.rs [hunk 3]
```diff
@@ -141,3 +141,26 @@ impl<SF: SubgraphServiceFactory> ExecutionServiceFactory for ExecutionCreator<SF
         ExecutionRequest,
     >>::Future;
 }
+
```
Comment (garypen) [inline apollo-router/src/services/execution_service.rs:144]: This is puzzling me. Why is this required as part of a change to the PluginTestHarness?

File: apollo-router/src/services/mod.rs (hunk 2 above)
Comment (SimonSapin) [reply apollo-router/src/services/mod.rs:125]: Done

File: apollo-router/src/services/execution_service.rs (hunk 3 above)
Comment (SimonSapin) [reply apollo-router/src/services/execution_service.rs:144]: This is required by a new feature of `PluginTestHarness`: it now has a `mock_execution_service` parameter, so that any stage of the pipeline can be mocked. I’ve added it to the changelog.

File: apollo-router/src/plugin/test/mod.rs [hunk 4]
```diff
@@ -257,15 +198,16 @@ mod testing {
...
     #[tokio::test]
     async fn test_test_harness() -> Result<(), BoxError> {
         let mut harness = PluginTestHarness::builder()
-            .plugin(EmptyPlugin {})
-            .schema(IntoSchema::Canned)
+            .configuration(json!({ "test.empty_plugin": null }))
             .build()
             .await?;
         let graphql = harness.call_canned().await?.next_response().await.unwrap();
+        assert_eq!(graphql.errors, []);
```
Comment (o0Ignition0o) [inline apollo-router/src/plugin/test/mod.rs:210]: nit
```suggestion
        assert!(graphql.errors.is_empty());
```

This imo would look more symetrical with the kinds of assertions [apollo-rs does](https://github.com/apollographql/apollo-rs/blob/233ed800417c00575c601722f5204bf91cace4be/crates/apollo-compiler/src/lib.rs#L211)

File: apollo-router/src/plugins/csrf.rs [hunk 5]
```diff
@@ -263,7 +263,10 @@ mod csrf_tests {
     #[tokio::test]
     async fn it_rejects_non_preflighted_headers_request() {
         let config = CSRFConfig::default();
-        let non_preflighted_request = RouterRequest::fake_builder().build().unwrap();
+        let non_preflighted_request = RouterRequest::fake_builder()
+            .header("content-type", "text/plain")
```
Comment (o0Ignition0o) [inline apollo-router/src/plugins/csrf.rs:267]: does the header() method replace or append a header?

Raising this because the changelog mentions `When no Content-Type header is specified, this builder will now default to application/json which makes the request be accepted by CSRF protection.` so I'd suspect the header value to now be ["application/json", "text/plain"] which would be kinda surprising?

File: apollo-router/src/plugins/telemetry/mod.rs [hunk 6]
```diff
@@ -174,8 +174,10 @@ impl Plugin for Telemetry {
         let telemetry = tracing_opentelemetry::layer().with_tracer(tracer);
         Self::replace_tracer_provider(tracer_provider);
 
-        replace_layer(Box::new(telemetry))
-            .expect("set_global_subscriber() was not called at startup, fatal");
+        let result = replace_layer(Box::new(telemetry));
+        if cfg!(not(test)) {
+            result.expect("set_global_subscriber() was not called at startup, fatal");
+        }
```
Comment (o0Ignition0o) [inline apollo-router/src/plugins/telemetry/mod.rs:180]: This might be overlapping with @Geal s work?

File: apollo-router/src/plugin/test/mod.rs [hunk 7]
```diff
@@ -207,6 +137,14 @@ impl PluginTestHarness {
             )
             .await
     }
+
+    /// Return the plugin instance that has the given `name`
+    ///
+    /// Returns `None` if the name or `Plugin` type does not match those given to [`register_plugin!`],
+    /// or if that plugin was not enabled in configuration.
+    pub fn plugin<Plugin: 'static>(&self, name: &str) -> Option<&Plugin> {
+        self.plugins.get(name)?.downcast_ref::<Plugin>()
```
Comment (o0Ignition0o) [inline apollo-router/src/plugin/test/mod.rs:146]: Oh this is why we have the downcast magic?

I see!

**Review by o0Ignition0o — COMMENTED** [review]
I like where this is going!

There's a few things I m not sure I understand, but it looks great overall!

File: apollo-router/src/plugin/mod.rs (hunk 1 above)
Comment (o0Ignition0o) [reply apollo-router/src/plugin/mod.rs]: ++ on the source url that would be nice ^^

File: apollo-router/src/plugins/telemetry/mod.rs (hunk 6 above)
Comment (Geal) [reply apollo-router/src/plugins/telemetry/mod.rs:180]: this should not be neccessary once #1463 lands. And with the current version of the code it smells like a workaround for a deeper issue :upside_down_face: 
so idk, either dig into it or wait for the fix :grin:

**Comment by SimonSapin** [comment]
After trying to follow this up by porting more things to use `PluginTestHarness` I’m now reconsidering this approach.

A lot of use cases involve manipulating a plugin instance. Already this PR does trait object downcasting to work around that, but having stuff in `Arc`s makes mutability more difficult.

Then I’m considering generalizing the harness API: instead of a `MockRouterService`, take something like `BoxService<RouterRequest, RouterResponse>`. But now that API starts looking a lot like the `Plugin` trait. So maybe the test harness should take a configuration *and* an (optional) plugin instance.

**Comment by SimonSapin** [comment]
I will soon submit a PR with a different approach that supersedes this PR.
